### ISSUE TITLE
ACK_MP can be received on any path

### ISSUE THREAD

**Issue by huitema** [quicwg/multipath#190]
The current text says that "ACK_MP frame (defined in {{ack-mp-frame}}) SHOULD be sent on (the same) path."  (The exact wording will change as part of PR #185 but the meaning will be the same.) The next sentence says "However, an ACK_MP frame can be returned via a
different path, based on different strategies of sending ACK_MP frames." 

There are two issues with that. First, as stated, this text is just implementation guidance: the second sentence in fact says that ACK_MP can be sent on any path, and that nodes MUST not treat receiving an ACK_MP on a "different" path as an error. If this is implementation guidance, the text should move from the normative section to the implementation guidance section.

Second, I believe the guidance is wrong. I understand that the "same path" practice makes path RTT computation easier, but that's only one aspect of the issue, IMHO not the most dominant aspect. In QUIC, most packet losses are detected by noticing that packets sent later on the path are already acknowledged, not by waiting on timers. Sending ACK_MP on a short path will enable faster detection of packet losses by acked number holes, while sending only on the "same" path only improve the computation of RTT, which may or may not improve timer-based recovery. So, this is a trade-off: more precise path-RTT computation versus shorter control loop. Such tradeoff belong to implementation guidance, not the normative part.

At some point, I contributed the algorithm used in Picoquic to derive path estimates through multi-dimensional statistical analysis. OK, I get the criticism that we should not require every implementer to have a PhD in applied maths. There are of course much simpler ways to do this. Suppose a packet is sent on path X and the ACK received on path Y. The RTT will be approximately 1/2 the RTT of path X plus 1/2 the RTT of path Y. If we assume that delay variations are dominated by effects on the send path, we can apply a simple transform:

1- let S be an RTT sample with data transmission on path X and ack on path Y.
2- assume that S = 1/2 RTT(X) + 1/2 RTT(Y)
3- derive S' = 2*S - RTT(Y)
4- consider S' as an approximation of a measurement of RTT(X)
5- use S' to update the smoothed RTT of path X, the RTT deviation, and the PTO.

We can discuss whether step 3 should use the smoothed RTT observed on path Y or the min RTT. Using the min RTT is probably more robust. The key here is that we can update the PTO of the path using very simple math.

But we can also wonder whether this is even needed. The PTO triggers when there is suspicion that the last packets sent to a path were lost. Effectively, the PTO triggers when "at this time, an ACK_MP should have been received". If ACK_MP are usually sent on the shortest path, it seems fine to just use the observed ACK delay (S in the example above) to update the PTO. The PTO will reflect both the delays and the ACK_MP sending strategy of the peer, and will fire exactly when the ACK_MP of the peer should most probably have been received.

And of course the value could end up being wrong -- but that's not a new issue. The statistical model behind the PTO computation in RFC 9002 is an approximation (see "no PhD" comment), and that too is wrong some of the time. That's why RFC 9002 specifies doubling the PTO delay after each trial, and also not changing congestion parameter simply because of a first PTO expiration.

Bottom line: mandating ACK_MP on the same path is a tradeoff between ease of path RTT computation and efficiency of loss recovery. It should not be mandated. If we want to provide guidance, that should be in the implementation section.

**Comment by LPardue** [comment]
Discussed in IETF 116: agreement that this is an issue but not whether it should be normative or not. Suggestion to start with adding more text to explain in more detail what happens when different choices are made and then later consider if this should have some normative mechanism

**Comment by ianswett** [comment]
The conversation in the room was about the importance of a consistent ACK strategy in order to provide a stable RTT for a given path.

RTT is used by PTO and some congestion controllers.  Changes in path characteristics do happen, so both PTO and congestion controllers should be designed to adapt to changes in strategy, but when those occur there's an increased chance of spurious or delayed PTOs as well as over or under-sending data into the network.

ACKs do not need to be sent on the same path they were received upon to provide a stable RTT and representative RTTVar, but the strategy should be stable when possible.  For example, even alternating paths on every ACK produces RTT and RTTVar values that should not cause spurious PTOs, because the RTT will be the mean RTT and the farther apart the RTTs are, the larger the RTTVar will be.

Looking at RFC9002, there are a number of normative statements about RTT calculation, and some of those were added to minimize the possibility of spurious PTOs, but there's nothing that parallels this I can find, so I could go either way on normative text.

One reason to send an ACK on the same path that data is received upon is to ensure network devices (ie: NATs, etc) keep the path open.  Sending in one direction on a path risks having the path get blackholed.

**Comment by huitema** [comment]
So the default strategy would be the "naive" strategy, in which every RTT sample contributes to smoother RTT and RTT VAR, regardless of the path the ACK was sent on. The smoothed RTT and RTT VAR will incorporate both the variations in path delays and the ACK strategy of the receiver. In general, acking through the "shortest" path results in shorter RTT, and thus more efficient loss recovery and congestion control. Using a stable strategy will tend to minimize RTT VAR, which also contributes to better performance.

**Comment by huitema** [comment]
And yes, there may be a need to send keep-alive traffic on all paths. But this is not quite the same issue as chosing paths to send ACKs.

**Comment by mirjak** [comment]
No commit was merged without PR. Sorry but fixed now.
